ffs/filcold/filcold.go

```diff
@@ -268,36 +269,52 @@ func (fc *FilCold) makeDeals(ctx context.Context, c cid.Cid, size uint64, cfgs [
 // If the deal finishes successfully it returns a FilStorage result.
 // If the deal finished with error, it returns a ffs.DealError error
 // result, so it should be considered in error handling.
-func (fc *FilCold) WaitForDeal(ctx context.Context, c cid.Cid, proposal cid.Cid) (ffs.FilStorage, error) {
+func (fc *FilCold) WaitForDeal(ctx context.Context, c cid.Cid, proposal cid.Cid, timeout time.Duration) (ffs.FilStorage, error) {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 	chDi, err := fc.dm.Watch(ctx, []cid.Cid{proposal})
 	if err != nil {
 		return ffs.FilStorage{}, fmt.Errorf("watching proposals in deals module: %s", err)
 	}
 
-	var activeProposal ffs.FilStorage
-	for di := range chDi {
-		switch di.StateID {
-		case storagemarket.StorageDealActive:
-			activeProposal = ffs.FilStorage{
-				ProposalCid:     di.ProposalCid,
-				PieceCid:        di.PieceCID,
-				Duration:        int64(di.Duration),
-				Miner:           di.Miner,
-				ActivationEpoch: di.ActivationEpoch,
-				StartEpoch:      di.StartEpoch,
-				EpochPrice:      di.PricePerEpoch,
+	var last deals.StorageDealInfo
+Loop:
+	for {
+		select {
+		case <-time.After(timeout):
+			msg := fmt.Sprintf("DealID %d with miner %s tracking timed out after waiting for %.0f hours.", last.DealID, last.Miner, timeout.Hours())
+			fc.l.Log(ctx, msg)
+			return ffs.FilStorage{}, ffs.DealError{ProposalCid: proposal, Message: msg}
+		case di, ok := <-chDi:
+			if !ok {
+				break Loop
+			}
+			last = di
+			switch di.StateID {
+			case storagemarket.StorageDealActive:
+				activeProposal := ffs.FilStorage{
+					ProposalCid:     di.ProposalCid,
+					PieceCid:        di.PieceCID,
+					Duration:        int64(di.Duration),
+					Miner:           di.Miner,
+					ActivationEpoch: di.ActivationEpoch,
+					StartEpoch:      di.StartEpoch,
+					EpochPrice:      di.PricePerEpoch,
+				}
+				fc.l.Log(ctx, "Deal %d with miner %s is active on-chain", di.DealID, di.Miner)
+				return activeProposal, nil
+			case storagemarket.StorageDealError, storagemarket.StorageDealFailing:
+				log.Errorf("deal %d failed with state %s: %s", di.DealID, storagemarket.DealStates[di.StateID], di.Message)
+				fc.l.Log(ctx, "DealID %d with miner %s failed and won't be active on-chain: %s", di.DealID, di.Miner, di.Message)
+
+				return ffs.FilStorage{}, ffs.DealError{ProposalCid: di.ProposalCid, Miner: di.Miner, Message: di.Message}
+			default:
+				if di.DealID != 0 {
+					fc.l.Log(ctx, "Deal %d with miner %s changed state to %s", di.DealID, di.Miner, storagemarket.DealStates[di.StateID])
+				} else {
+					fc.l.Log(ctx, "Deal with miner %s changed state to %s", di.Miner, storagemarket.DealStates[di.StateID])
+				}
```

Comment:
Before we printed the Deal ID only when the deal reached success.
Now we print it in any intermediary status that already has a value > 0, which usually is a status where the deal is on-chain but not yet active.

This is useful for debugging stuff, or talking with the miner asking: "Hey, what is happening with Deal X? Is still sealing". Since before this ID was only printed on success, it was a bit hard to do this.